## IT2313 - Programming for Data Science
## Practical 5 - Question 1 to 11
## Data Manipulation with Pandas (Part 1)

### In this lesson, we will continue to practise our Pandas knowledge. 
- Importing Data into Pandas
- Changing the Values in DataFrame
- DataFrameData Exploration, Sorting and Filtering
- Slicing and Subsetting with .loc and .iloc
- Aggregating Methods

#### We need to import the neccessary libraries

In [108]:
import numpy as np
import pandas as pd

#### Q1 - Let use the following Dictionary known as "dict" and load it into the Pandas DataFrame known as brics

In [109]:
dict = { "country":["Brazil", "Russia", "India", "China", "South Africa"],
         "capital":["Brasilia", "Moscow", "New Delhi", "Beijing", "Pretoria"],
         "area":[8.516, 17.10, 3.286, 9.597, 1.221],
         "population":[200.4, 143.5, 1252, 1357, 52.98] }
dict

{'country': ['Brazil', 'Russia', 'India', 'China', 'South Africa'],
 'capital': ['Brasilia', 'Moscow', 'New Delhi', 'Beijing', 'Pretoria'],
 'area': [8.516, 17.1, 3.286, 9.597, 1.221],
 'population': [200.4, 143.5, 1252, 1357, 52.98]}

In [110]:
# Type your Q1 answer here
brics=pd.DataFrame(dict)
brics


country    capital    area  population
0        Brazil   Brasilia   8.516      200.40
1        Russia     Moscow  17.100      143.50
2         India  New Delhi   3.286     1252.00
3         China    Beijing   9.597     1357.00
4  South Africa   Pretoria   1.221       52.98

#### Q2 - Change the Dataframe Index to : ["BR", "RU", "IN", "CH", "SA"]

In [111]:
# Type your Q2 answer here

index=["BR", "RU", "IN", "CH", "SA"]
brics=pd.DataFrame(dict,index=index)
brics

country    capital    area  population
BR        Brazil   Brasilia   8.516      200.40
RU        Russia     Moscow  17.100      143.50
IN         India  New Delhi   3.286     1252.00
CH         China    Beijing   9.597     1357.00
SA  South Africa   Pretoria   1.221       52.98

#### Q3 - Let export your DataFrame into a csv file know as brics.csv. Use Excel to open the csv file and compare with the DataFrame

In [112]:
# Type your Q3 answer here
# Export DataFrame to CSV
brics.to_csv('brics.csv')


#### Q4 - Let select the information for China using "loc" and display in Pandas Series and Pandas DataFrame

In [113]:
# Type your Q4 answer here for Pandas Series using "loc"
brics.loc["CH"]
 

country         China
capital       Beijing
area            9.597
population     1357.0
Name: CH, dtype: object

In [114]:
# Type your Q4 answer here for Pandas DataFrame using "loc"
brics.loc[["CH"]]


country  capital   area  population
CH   China  Beijing  9.597      1357.0

#### Q5 - Do select and display Row 2 to 4 of the DataFrame as shown

|    | country | capital   | area   | population |
|----|---------|-----------|--------|------------|
| RU | Russia  | Moscow    | 17.100 | 143.5      |
| IN | India   | New Delhi | 3.286  | 1252.0     |
| CH | China   | Beijing   | 9.597  | 1357.0     |

In [115]:
# Type your Q5 answer here, there are many ways to achieve the selection.
print(brics.iloc[1:4])
print(brics[1:4])


   country    capital    area  population
RU  Russia     Moscow  17.100       143.5
IN   India  New Delhi   3.286      1252.0
CH   China    Beijing   9.597      1357.0
   country    capital    area  population
RU  Russia     Moscow  17.100       143.5
IN   India  New Delhi   3.286      1252.0
CH   China    Beijing   9.597      1357.0


#### Q6 - Let use "loc" to select the and display the following:
|    | country | capital   | area   | population |
|----|---------|-----------|--------|------------|
| RU | Russia  | Moscow    | 17.100 | 143.5      |
| IN | India   | New Delhi | 3.286  | 1252.0     |

In [116]:
# Type your Q6 answer here
brics.loc["RU":"IN"]


country    capital    area  population
RU  Russia     Moscow  17.100       143.5
IN   India  New Delhi   3.286      1252.0

In [117]:
brics[1:4]

country    capital    area  population
RU  Russia     Moscow  17.100       143.5
IN   India  New Delhi   3.286      1252.0
CH   China    Beijing   9.597      1357.0

#### Q7 - Let use "loc" to select the and display the following:
|    | country | capital   |
|----|---------|-----------|
| RU | Russia  | Moscow    |
| IN | India   | New Delhi |

In [118]:
# Type your Q7 answer here
brics.loc["RU":"IN", "country":"capital"]


country    capital
RU  Russia     Moscow
IN   India  New Delhi

#### Q8 - Let use "loc" to select the and display the following:
|    | country       | capital           |
|----|---------------|-------------------|
| BR | Brazil        | Brasilia          |
| RU | Russia        | Moscow            |
| IN | India         | New Delhi         |
| CH | China         | Beijing           |
| SA | South Africia | Pretoria          |

In [119]:
# Type your Q8 answer here
brics.loc[:,"country":"capital"]


country    capital
BR        Brazil   Brasilia
RU        Russia     Moscow
IN         India  New Delhi
CH         China    Beijing
SA  South Africa   Pretoria

#### Q9 - Let use "iloc" to select the and display the following:

|    | country | capital   | area   | population |
|----|---------|-----------|--------|------------|
| RU | Russia  | Moscow    | 17.100 | 143.5      |
| IN | India   | New Delhi | 3.286  | 1252.0     |
| CH | China   | Beijing   | 9.597  | 1357.0     |

In [120]:
# Type your Q9 answer here
brics.iloc[1:4]


country    capital    area  population
RU  Russia     Moscow  17.100       143.5
IN   India  New Delhi   3.286      1252.0
CH   China    Beijing   9.597      1357.0

#### Q10 - Let use "iloc" to select the and display the following:

|    | country | capital   |
|----|---------|-----------|
| RU | Russia  | Moscow    |
| IN | India   | New Delhi |
| CH | China   | Beijing   |

In [121]:
# Type your Q10 answer here
brics.iloc[:,0:2]


country    capital
BR        Brazil   Brasilia
RU        Russia     Moscow
IN         India  New Delhi
CH         China    Beijing
SA  South Africa   Pretoria

##### Q11 - Let import dataset from "temperatures.csv" into DataFrame known as temperatures

- a. Display the top 5 and bottom 5 rows of the DataFrame
- b. Display the number of rows and columns in this DataFrame
- c. Drop the column "s/n" from the DataFrame
- d. Generate descriptive statistics of the DataFrame's numerical columns
- e. Setting the index of the DataFrame to be a multi-level index using the columns "country" and "city" and display the Top 5 rows
- f. Sort the DataFrame based on Indexes in _ascending_ order and display the Top 5 rows and display the top 5 rows.
- g. Select rows starting from the combination ("India", "Hyderabad") up to and including the combination ("Iraq", "Baghdad").
- h. Select the rows which filter the rows in the DataFrame based on the specified date range. From 1 Jan 2010 to 31 Jan 2010

In [122]:
# Import your Dataset from "temperatures.csv" into DataFrame known as temperatures
temperatures=pd.read_csv('temperatures.csv')

In [123]:
# Type your Q11 (a) answer here
temperatures.head(5)


s/n      date     city        country  avg_temp_c
0    0  1/1/2000  Abidjan  Côte D'Ivoire      27.293
1    1  1/2/2000  Abidjan  Côte D'Ivoire      27.685
2    2  1/3/2000  Abidjan  Côte D'Ivoire      29.061
3    3  1/4/2000  Abidjan  Côte D'Ivoire      28.162
4    4  1/5/2000  Abidjan  Côte D'Ivoire      27.547

In [124]:
# Type your Q11 (a) answer here
temperatures.tail(5)


s/n      date  city country  avg_temp_c
16495  16495  1/5/2013  Xian   China      18.979
16496  16496  1/6/2013  Xian   China      23.522
16497  16497  1/7/2013  Xian   China      25.251
16498  16498  1/8/2013  Xian   China      24.528
16499  16499  1/9/2013  Xian   China         NaN

In [125]:
# Type your Q11 (b) answer here
print(temperatures.shape)

(16500, 5)


In [126]:
# Type your Q11(c) answer here
temperatures=temperatures.drop(axis=1,columns="s/n")
temperatures.head()


date     city        country  avg_temp_c
0  1/1/2000  Abidjan  Côte D'Ivoire      27.293
1  1/2/2000  Abidjan  Côte D'Ivoire      27.685
2  1/3/2000  Abidjan  Côte D'Ivoire      29.061
3  1/4/2000  Abidjan  Côte D'Ivoire      28.162
4  1/5/2000  Abidjan  Côte D'Ivoire      27.547

In [127]:
# Type your Q11(d) answer here
temperatures.describe()


avg_temp_c
count  16407.000000
mean      19.732877
std        9.579635
min      -23.495000
25%       14.870000
50%       22.119000
75%       27.001500
max       38.283000

In [128]:
temperatures.columns.nlevels

1

In [132]:
# Type your Q11(e) answer here
temperatures_indx=temperatures.set_index(["country","city"])
temperatures_indx.head()

date  avg_temp_c
country       city                          
Côte D'Ivoire Abidjan 2000-01-01      27.293
              Abidjan 2000-01-02      27.685
              Abidjan 2000-01-03      29.061
              Abidjan 2000-01-04      28.162
              Abidjan 2000-01-05      27.547

In [134]:
# Type your Q11(f) answer here
temperatures_srt=temperatures_indx.sort_index()
temperatures_srt

date  avg_temp_c
country     city                         
Afghanistan Kabul  2000-01-01       3.326
            Kabul  2000-01-02       3.454
            Kabul  2000-01-03       9.612
            Kabul  2000-01-04      17.925
            Kabul  2000-01-05      24.658
...                       ...         ...
Zimbabwe    Harare 2013-01-05      18.298
            Harare 2013-01-06      17.020
            Harare 2013-01-07      16.299
            Harare 2013-01-08      19.232
            Harare 2013-01-09         NaN

[16500 rows x 2 columns]

In [135]:
# Type your Q11(g) answer here

temperatures_srt.loc[("India","Hyderabad"):("Iraq","Baghdad")]

date  avg_temp_c
country city                            
India   Hyderabad 2000-01-01      23.779
        Hyderabad 2000-01-02      25.826
        Hyderabad 2000-01-03      28.821
        Hyderabad 2000-01-04      32.698
        Hyderabad 2000-01-05      32.438
...                      ...         ...
Iraq    Baghdad   2013-01-05      28.673
        Baghdad   2013-01-06      33.803
        Baghdad   2013-01-07      36.392
        Baghdad   2013-01-08      35.463
        Baghdad   2013-01-09         NaN

[2145 rows x 2 columns]

In [130]:
# Type your Q11(h) answer here
print(temperatures)
temperatures["date"]=pd.to_datetime(temperatures["date"],format="%m/%d/%Y")
temperatures

           date     city        country  avg_temp_c
0      1/1/2000  Abidjan  Côte D'Ivoire      27.293
1      1/2/2000  Abidjan  Côte D'Ivoire      27.685
2      1/3/2000  Abidjan  Côte D'Ivoire      29.061
3      1/4/2000  Abidjan  Côte D'Ivoire      28.162
4      1/5/2000  Abidjan  Côte D'Ivoire      27.547
...         ...      ...            ...         ...
16495  1/5/2013     Xian          China      18.979
16496  1/6/2013     Xian          China      23.522
16497  1/7/2013     Xian          China      25.251
16498  1/8/2013     Xian          China      24.528
16499  1/9/2013     Xian          China         NaN

[16500 rows x 4 columns]


date     city        country  avg_temp_c
0     2000-01-01  Abidjan  Côte D'Ivoire      27.293
1     2000-01-02  Abidjan  Côte D'Ivoire      27.685
2     2000-01-03  Abidjan  Côte D'Ivoire      29.061
3     2000-01-04  Abidjan  Côte D'Ivoire      28.162
4     2000-01-05  Abidjan  Côte D'Ivoire      27.547
...          ...      ...            ...         ...
16495 2013-01-05     Xian          China      18.979
16496 2013-01-06     Xian          China      23.522
16497 2013-01-07     Xian          China      25.251
16498 2013-01-08     Xian          China      24.528
16499 2013-01-09     Xian          China         NaN

[16500 rows x 4 columns]

In [131]:
temperatures_bool = temperatures[
    (temperatures["date"] >= "2010-01-01") &
    (temperatures["date"] <= "2010-01-31")
]
temperatures_bool

date     city        country  avg_temp_c
120   2010-01-01  Abidjan  Côte D'Ivoire      28.270
121   2010-01-02  Abidjan  Côte D'Ivoire      29.262
122   2010-01-03  Abidjan  Côte D'Ivoire      29.596
123   2010-01-04  Abidjan  Côte D'Ivoire      29.068
124   2010-01-05  Abidjan  Côte D'Ivoire      28.258
...          ...      ...            ...         ...
16462 2010-01-08     Xian          China      23.434
16463 2010-01-09     Xian          China      19.630
16464 2010-01-10     Xian          China      12.292
16465 2010-01-11     Xian          China       6.742
16466 2010-01-12     Xian          China       0.845

[1200 rows x 4 columns]